custom_from.php

```diff
@@ -14,63 +14,177 @@
         $this->add_hook('message_compose_body', array($this, 'message_compose_body'));
         $this->add_hook('render_page', array($this, 'render_page'));
         $this->add_hook('storage_init', array($this, 'storage_init'));
+        $this->add_hook('preferences_sections_list', array($this, 'preferences_sections_list'));
+        $this->add_hook('preferences_list', array($this, 'preferences_list'));
+        $this->add_hook('preferences_save', array($this, 'preferences_save'));
+        $this->add_hook('template_object_composeheaders', array($this, 'template_object_composeheaders'));
+        $this->add_hook('template_object_composebody', array($this, 'template_object_composebody'));
     }
 
     /**
-     ** Adds additional headers to supported headers list.
+     ** Override form fields
      */
-    public function storage_init($params)
+    public function template_object_composeheaders($attrib)
     {
         $this->load_config();
 
         $rcmail = rcmail::get_instance();
-        $rules = self::get_rules($rcmail->config);
+        $compose_id = rcube_utils::get_input_value('_id', rcube_utils::INPUT_GET);
+        $compose_data = $_SESSION['compose_data_' . $compose_id] ?? array();
+        $msg_uid = '';
+
+        if (isset($compose_data['param']['reply_uid'])) {
+            $msg_uid = $compose_data['param']['reply_uid'];
+            self::$is_reply = true;
+        } elseif (isset($compose_data['param']['draft_uid'])) {
+            $msg_uid = $compose_data['param']['draft_uid'];
+            self::$is_draft = true;
+        } elseif (isset($compose_data['param']['forward_uid'])) {
+            $msg_uid = $compose_data['param']['forward_uid'];
+        } elseif (isset($compose_data['param']['uid'])) {
+            $msg_uid = $compose_data['param']['uid'];
+        }
 
-        foreach (array_keys($rules) as $header) {
-            $params['fetch_headers'] = (isset($params['fetch_headers']) && $params['fetch_headers'] !== '' ? $params['fetch_headers'] . ' ' : '') . $header;
+        if (!$rcmail->config->get('custom_from_compose_auto', true) || $msg_uid === '') {
+            return $attrib;
         }
 
-        return $params;
+        $reply_from = rcmail::get_instance()->user->prefs[self::REPLY_FROM_SETTING] ?? self::RECEIVING_EMAIL_WITH_IDENTITY_OPTION;
+
+        switch ($attrib['part']) {
+            case 'from':
+                if (self::$is_draft) {
+                    $attrib = $this->compose_draft_from_headers($compose_id, $msg_uid, $reply_from, $attrib);
+                } else {
+                    $attrib = $this->compose_from_headers($compose_id, $msg_uid, $reply_from, $attrib);
+                }
+                break;
+            case 'replyto':
+            case 'bcc':
+                if (!self::$is_draft) {
+                    $attrib = $this->compose_additional_headers($compose_id, $msg_uid, $reply_from, $attrib);
+                }
+            default:
+                break;
```

Comment:
No-op?

Reply:
The methods `compose_additional_headers`, `compose_draft_from_headers`, and `compose_from_headers` are triggered if there is already a saved or received message.

In the `compose_draft_from_headers` method, we add the ability to load a custom sender into the text field in drafts.
In the `compose_from_headers` method, we replace the sender if necessary.

With the option `self::RECEIVING_EMAIL_OPTION`, we reply from the email to which the message was received.
With the option `self::RECEIVING_EMAIL_WITH_IDENTITY_OPTION`, we reply from the email to which the message was received and use the IDENTITY settings for the other fields.
With the option `self::RECEIVING_EMAIL_WITH_DEFAULT_IDENTITY_OPTION`, we reply from the default identity, and we replace the select field to leave only the default identity in the reply field since the Roundcube JS functionality does not allow otherwise.

In the `compose_additional_headers` method, we clear or replace fields for two options. Roundcube will independently fill in the fields on the frontend for other options. For our cases, we need to use replacements with our own fields so that the JS functionality does not overwrite our values.

Initially, the field value is set to empty and with the `self::RECEIVING_EMAIL_OPTION` option it is simply cleared and replaced with our `html_textarea`. With the `self::RECEIVING_EMAIL_WITH_DEFAULT_IDENTITY_OPTION` option, we also create our own fields so that the Roundcube JS functionality does not overwrite them, and fill the fields with values from the default identity.